examples/cpp/keyvaluestore/caching_interceptor.h

```diff
@@ -0,0 +1,128 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <map>
+
+#include <grpcpp/support/client_interceptor.h>
+
+#ifdef BAZEL_BUILD
+#include "examples/protos/keyvaluestore.grpc.pb.h"
+#else
+#include "keyvaluestore.grpc.pb.h"
+#endif
+
+// This is a naive implementation of a cache. A new cache is for each call. For
+// each new key request, the key is first searched in the map and if found. Only
+// if the key is not found in the cache do we make a request.
+class CachingInterceptor : public grpc::experimental::Interceptor {
+ public:
+  CachingInterceptor(grpc::experimental::ClientRpcInfo* info) {}
+
+  void Intercept(
+      ::grpc::experimental::InterceptorBatchMethods* methods) override {
+    bool hijack = false;
+    if (methods->QueryInterceptionHookPoint(
+            grpc::experimental::InterceptionHookPoints::
+                PRE_SEND_INITIAL_METADATA)) {
+      // Hijack all calls
+      hijack = true;
+      // Create a stream on which this interceptor can make requests
+      stub_ = keyvaluestore::KeyValueStore::NewStub(
+          methods->GetInterceptedChannel());
+      stream_ = stub_->GetValues(&context_);
+    }
+    if (methods->QueryInterceptionHookPoint(
+            grpc::experimental::InterceptionHookPoints::PRE_SEND_MESSAGE)) {
+      // We know that clients perform a Read and a Write in a loop, so we don't
+      // need to maintain a list of the responses.
+      std::string requested_key;
+      const keyvaluestore::Request* req_msg =
+          static_cast<const keyvaluestore::Request*>(methods->GetSendMessage());
+      if (req_msg != nullptr) {
+        requested_key = req_msg->key();
+      } else {
+        // The non-serialized form would not be available in certain scenarios,
```

Comment:
Could you add an example of the scenarios in which the non-serialized form is not available? E.g., "such as the async api"?

Reply:
saying that it is not available for async api is not always true. The most accurate description is at https://github.com/grpc/grpc/pull/17662/files